src/Frame.cc

```diff
@@ -1385,57 +1379,257 @@ void Frame::SaveFile(const std::string& root_folder, const CARTA::SaveFile& save
+            }
+        } else {
+            return;
         }
 
-        // Get a copy of the original pixel data
-        casacore::IPosition start(image->shape().size(), 0);
-        casacore::IPosition count(image->shape());
-        casacore::Slicer slice(start, count);
-        casacore::Array<casacore::Float> temp_array;
-        image->doGetSlice(temp_array, slice);
+        bool is_keep_degenerate = false;
+        if (save_file_msg.keep_degenerate()) {
+            is_keep_degenerate = true;
+        }
+        auto slice_sub_image = casacore::Slicer(start, end, stride, casacore::Slicer::endIsLast);
+        casacore::SubImage<float> sub_image;
 
-        // Construct a new CASA image
         try {
-            auto out_image =
-                std::make_unique<casacore::PagedImage<casacore::Float>>(image->shape(), image->coordinates(), output_filename.string());
-            out_image->setMiscInfo(image->miscInfo());
-            out_image->setImageInfo(image->imageInfo());
-            out_image->appendLog(image->logger());
-            out_image->setUnits(image->units());
-            out_image->putSlice(temp_array, start);
-
-            // Copy the mask if the original image has
-            if (image->hasPixelMask()) {
-                casacore::Array<casacore::Bool> image_mask;
-                image->getMaskSlice(image_mask, slice);
-                out_image->makeMask("mask0", true, true);
-                casacore::Lattice<casacore::Bool>& out_image_mask = out_image->pixelMask();
-                out_image_mask.putSlice(image_mask, start);
-            }
+            _loader->GetSubImage(slice_sub_image, sub_image, is_keep_degenerate);
+            image = sub_image.cloneII();
         } catch (casacore::AipsError error) {
             message = error.getMesg();
             save_file_ack.set_success(false);
             save_file_ack.set_message(message);
             return;
         }
-        success = true;
-    } else if (output_file_type == CARTA::FileType::FITS) {
-        // Remove the old image file if it has a same file name
-        casacore::Bool ok = casacore::ImageFITSConverter::ImageToFITS(message, *image, output_filename.string(), 64, casacore::True,
-            casacore::True, -32, 1.0, -1.0, casacore::True, casacore::False, casacore::True, casacore::False, casacore::False,
-            casacore::False, casacore::String(), casacore::True);
-        if (ok) {
-            success = true;
+    }
+
+    try {
+        std::unique_lock<std::mutex> ulock(_image_mutex); // Lock the image while saving the file
+        switch (output_file_type) {
+            case CARTA::FileType::CASA:
+                success = ExportCASAImage(*image, output_filename, message);
+                break;
+            case CARTA::FileType::FITS:
+                success = ExportFITSImage(*image, output_filename, message);
+                break;
+            default:
+                message = "No saving file action!";
```

Comment:
This should be something like "Could not export file. Unknown file type", with the type appended.

Reply:
Replaced